actioncable/lib/action_cable/server/connection.rb

```diff
@@ -0,0 +1,198 @@
+# frozen_string_literal: true
+
+require "action_dispatch"
+
+module ActionCable
+  module Server
+    # This class encapsulates all the low-level logic of working with the underlying WebSocket conenctions
+    # and delegate all the business-logic to the user-level connection object (e.g., ApplicationCable::Connection).
+    # This connection object is also responsible for handling encoding and decoding of messages, so the user-level
+    # connection object shouldn't know about such details.
+    class Connection
+      attr_reader :server, :env, :protocol, :logger, :app_conn
+      private attr_reader :worker_pool
+
+      delegate :event_loop, :pubsub, :config, to: :server
+
+      def initialize(server, env, coder: ActiveSupport::JSON)
+        @server, @env, @coder = server, env, coder
+
+        @worker_pool = server.worker_pool
+        @logger = new_tagged_logger
+
+        @websocket      = Server::Connection::WebSocket.new(env, self, event_loop)
+        @message_buffer = Server::Connection::MessageBuffer.new(self)
+
+        @protocol = nil
+        @app_conn = config.connection_class.call.new(server, self)
```

Comment:
I wonder if we should cache the result of `config.connection_class.call`?

Reply:
I guess, the primary reason why it's implemented like that is code reloading. We can check the app configuration value and decide if we need to cache it or not; however, I'm not sure it worth the effort.

Also, I see a potential extension here—we can turn it into a connection class selector depending on the request information:

```ruby
config.connection_class = ->(request) {
  if request.path == "/admin/cable"
    Admin:Connection
  else
    ApplicationCable::Connection
  end 
}
```

So, keeping it dynamic makes sense.

Reply:
This is useful, I'm working on a NoStr server, I hack a `NostrConnection`, and inject to ActionCable in this way

```ruby
ActionCable.module_eval do
  module_function def nostr_server
    @nostr_server ||=
      begin
        config = ActionCable::Server::Base.config.dup
        config.connection_class = -> { "NostrCable::Connection".safe_constantize }
        ActionCable::Server::Base.new config: config
      end
  end
end
```